Add support for 'msg' and 'block'

djmgit opened the pull request:
Added support for special commands like 'msg' and
'block'. Return a struct containing all the
required data.

Fixes issue #7

djmgit commented:
![screenshot from 2018-06-21 21-54-21](https://user-images.githubusercontent.com/16368427/41732641-251e0004-759f-11e8-826a-c91d1bdcd033.png)

djmgit commented:
@raineorshine please have a look.

raineorshine commented on index.js:
```diff
@@ -130,6 +157,14 @@ module.exports = options => {
+    if (Object.keys(specialCommands).indexOf(commandWithoutSemi) !== -1) {
+      return evalSpecialSol(commandWithoutSemi, commands)
+        .then(result => {
+          return JSON.stringify(result)
```
I would suggest changing this to `JSON.stringify(result, null, 2)` to pretty print the result.

raineorshine commented on index.js:
```diff
@@ -117,6 +123,27 @@ ${source.split('\n').map((line, n) => (n + 1) + '  ' + line).join('\n')}
+  return new Bluebird((resolve, reject) => {
+    props.forEach(prop => {
+    let statement = command + "." + prop + ";"
+    evalSol(commands.concat(statement), options)
```
The `evalSol` command is quite slow, so I don't think it's viable to call it once for each `prop` for performance reasons.

I would suggest evaluating in a single, combined command, returning a tuple, and then unpacking the tuple into a readable struct.

This should also allow us to reuse `evalSol` without another version of it.

raineorshine commented on index.js:
```diff
@@ -16,6 +16,12 @@ const pkg = require('./package.json')
+const specialCommands = {
+  "msg": ["data", "gas", "sender", "sig", "value"],
+  "block": ["coinbase", "difficulty", "gaslimit", "number", "timestamp"]
+}
```
Missing:

- `block.blockhash`
- `tx.gasprice`
- `tx.origin`

raineorshine reviewed (changes requested):
Thanks for the initial PR! Let me know if you have any questions about my suggestions.

djmgit commented:
@raineorshine I have updated my PR. Please have a look.

![screenshot from 2018-06-25 01-13-49](https://user-images.githubusercontent.com/16368427/41822989-d67a244a-7815-11e8-8836-61ab65e91109.png)

block.blockhash is giving the following error:

```
> block.blockhash
Error compiling Solidity:

1  pragma solidity ^0.4.19;
2  contract ReplContainer {
3    function Main() constant returns (function (uint256) view returns (bytes32)) {
4      
5      return block.blockhash;
6    }
7  }

  :3:37: TypeError: Internal or recursive type is not allowed for public or external functions.
  function Main() constant returns (function (uint256) view returns (bytes32)) {
                                    ^----------------------------------------^
 :3:37: TypeError: Internal or recursive type is not allowed for public or external functions.
  function Main() constant returns (function (uint256) view returns (bytes32)) {
                                    ^----------------------------------------^

Error compiling Solidity
```

raineorshine commented:
Thanks! I'll take a look. I can see now that `block.blockhash` has been deprecated, so you can ignore that one.

raineorshine commented on index.js:
```diff
@@ -39,16 +62,41 @@ const getAccounts = () => {
   /** Returns true if the given command is an expression that can return a value. */
   const isExpression = command => (!/[^=]=[^=]/.test(command)) && (!command.startsWith('delete'))
-  const lastCommand = commands[commands.length - 1]
+  let lastCommand = commands[commands.length - 1]
```
It is recommended to keep `lastCommand` a `const` since it is never intended to be mutated.

raineorshine commented on index.js:
```diff
@@ -39,16 +62,41 @@ const getAccounts = () => {
+  let contentWithReturnExpression = commands.join('\n')
+
+  let sourceFirstPass = template({ content: contentWithReturnExpression, returnType, returnExpression: isExpression(lastCommand) ? lastCommand : 'false;' })
+  if (Object.keys(specialCommands).indexOf(commandWithoutSemi) !== -1) {
```
This can be clearer and more efficient using a single assignment and ternary operator:

```
const specialCommand = specialCommands[commandWithoutSemi]
const sourceFirstPass = template(specialCommand ? {
  content, 
  returnType: specialCommand.returnType,
  returnExpression: specialCommand.returnExpression
} : { 
  content: contentWithReturnExpression, 
  returnType: 'bool', 
  returnExpression: isExpression(lastCommand) ? lastCommand : 'false;' 
})
```

raineorshine commented on index.js:
```diff
@@ -131,9 +179,19 @@ module.exports = options => {
     }
 
     const commandWithSemi = command + (command.endsWith(';') ? '' : ';')
+    const commandWithoutSemi = command.endsWith(';') ? command.substring(0, command.length - 1) : command
```
Let's change the declaration on line 174 to:
```
const commandTrimmed = rawCommand.trim()
const command = commandTrimmed.endsWith(';') ? 
  commandTrimmed.substring(0, commandTrimmed.length - 1) : 
  commandTrimmed
```

Then we can simply use `command` later on and `command + ';'` where needed.

raineorshine commented on index.js:
```diff
@@ -131,9 +179,19 @@ module.exports = options => {
     return evalSol(commands.concat(commandWithSemi), options)
       .then(result => {
         commands.push(commandWithSemi)
+        if (Object.keys(specialCommands).indexOf(commandWithoutSemi) !== -1) {
```
A better condition would be `commandWithoutSemi in specialCommands`.

raineorshine commented on index.js:
```diff
@@ -16,6 +16,29 @@ const pkg = require('./package.json')
 const main = 'Main'
 
+// speacial commands
+const specialCommands = {
```
Let's rename this to `globalVars` or `specialGlobals`. I think that will be a little clearer.

raineorshine commented on index.js:
```diff
@@ -131,9 +179,19 @@ module.exports = options => {
         commands.push(commandWithSemi)
+        if (Object.keys(specialCommands).indexOf(commandWithoutSemi) !== -1) {
+          let struct = {}
+          let results = result.toString().split(",")
```
I believe `result` is always a string, so `.toString` is not needed. If it's not, we should make it clearer what the different types are and how they are handled.

raineorshine commented on index.js:
```diff
@@ -131,9 +179,19 @@ module.exports = options => {
+          let results = result.toString().split(",")
+          let props = specialCommands[commandWithoutSemi]
+          for (let i = 0; i < props.length; i++) {
+            struct[props[i].prop] = results[i]
```
Another way to do this would be with `reduce` instead of a `for` loop, but I don't have a strong preference here.

raineorshine reviewed (changes requested):
Please review comments for recommendations.

The only other thing that we will need is a unit test for this new feature.

Thanks!

djmgit commented:
@raineorshine I will be updating my PR ASAP.

djmgit commented:
@raineorshine I have tried to address all of your suggestions, please have a look.

raineorshine commented on index.js:
```diff
@@ -39,16 +62,41 @@ const getAccounts = () => {
+  let returnType = ""
+  let returnExpression = ""
+
+  const pluck = (arr, key) => arr.map(item => item[key])
```
Let's move this helper function up to the global scope for organizational purposes.

raineorshine commented on index.js:
```diff
@@ -39,16 +62,41 @@ const getAccounts = () => {
+
+  returnType = pluck(specialGlobals[command], "returnType").join(", ")
+  returnExpression = pluck(specialGlobals[command], "prop")
+  returnExpression = "(" + returnExpression.map(item => command + "." + item).join(", ") + ");"
```
Great, this is the right approach! Now, we can remove the function completely and just use the pluck + join expressions below where `getSpecialTemplateParams` is called. That eschews the extra function and keeps things cleaner since it is a small amount of code.

raineorshine commented on index.js:
```diff
@@ -39,16 +62,41 @@ const getAccounts = () => {
+  const sourceFirstPass = template(specialCommand ? {
+    content,
+    returnType: getSpecialTemplateParams(commandWithoutSemi).returnType,
+    returnExpression: getSpecialTemplateParams(commandWithoutSemi).returnExpression
```
Here instead of calling `getSpecialTemplateParams`, just inline the `pluck` + `join`.

raineorshine commented on index.js:
```diff
@@ -123,17 +171,30 @@ module.exports = options => {
+          let struct = {}
+
+          // result is an object, so we first convert it to string and then split
+          let results = result.toString().split(",")
```
Is the `.toString` actually needed here?

raineorshine replied:
It is not enough to know that it breaks without `toString`, we need to know *why* it breaks. `toString` is not a solution; it is rarely correct to use `toString` in Javascript. We need to understand what type `result` is and act on the object appropriately, deconstructing it if necessary. Use `typeof` and `instanceof` during debugging to attempt to determine what `result` actually is and how we can make an array of it. I would almost think `result` is already an array if we have to split the `toString` output on commas again.

raineorshine reviewed (changes requested):
Thank you! Just some minor changes left.

Please note however that this PR will not be merged without a unit test.

djmgit commented:
@raineorshine I have done the required changes. I have also added tests. Please have a look.
If I remove ```.toString``` from there, then its throwing error.

raineorshine commented on index.js:
```diff
@@ -44,11 +70,23 @@ const evalSol = (commands, options={}) => {
+    returnExpression: "(" + pluck(specialGlobals[commandWithoutSemi], "prop").map(item => commandWithoutSemi + "." + item).join(", ") + ");"
+  }: {
+    content: contentWithReturnExpression,
+    returnType,
```
I believe `returnType` in this case will always be `bool`. No need for a local variable when the value can be inlined.

raineorshine commented on test/test.js:
```diff
@@ -97,4 +97,37 @@ describe('solidity-repl', () => {
+  it('should execute special command msg', () => {
+    return repl('msg')
+      .then(result => {
+        assert.equal(Object.keys(JSON.parse(result)).length, 5)
```
I recommend adding `const parsedResult = JSON.parse(result)` or similar to avoid parsing the result more than once.

raineorshine commented on index.js:
```diff
@@ -44,11 +70,23 @@ const evalSol = (commands, options={}) => {
-  const sourceFirstPass = template({ content: contentWithReturnExpression, returnType, returnExpression: isExpression(lastCommand) ? lastCommand : 'false;' })
+  let contentWithReturnExpression = commands.join('\n')
+
+  //let sourceFirstPass = template({ content: contentWithReturnExpression, returnType, returnExpression: isExpression(lastCommand) ? lastCommand : 'false;' })
```
You can remove this comment.

raineorshine reviewed (changes requested):
Thank you for your work and for adding the unit tests. Just a couple minor points and we should be good to go. 

Try to notice the general rules you can infer infer from my feedback so that you can learn and apply it to all similar situations.

Thank you!

raineorshine reviewed (changes requested):
Added `toString` request.

djmgit committed:
Add support for 'msg' and 'block'

Added support for special commands like 'msg' and
'block'. Return a struct containing all the
required data.

Fixes issue #7

djmgit commented:
@raineorshine I was wrong about the result thing, there was no need for .toString as result is already an array, I ignored the fact that the typeof command returns object for arrays too!
I have removed the local variable ```returnType``` and have used the inline declarations in both places.
Please have a look.

raineorshine commented:
Thank you!
